### ISSUE TITLE
Show capacity for vertical sinking kernel

### ISSUE THREAD

**Issue by erikvansebille** [OceanPARCELS/parcels#13]
For the GMD manuscript, we will need to show that we can have particles sink at constant speed, as was done in e.g. http://www.nature.com/ncomms/2015/150304/ncomms7521/full/ncomms7521.html

**Comment by mlange05** [comment]
This feature would be a nice opportunity to implement and demonstrate kernel concatenation. This would allow the user to build custom kernels and combine them with existing ones that Parcels provides (eg. `AdvectionRK4`). The proposed syntax for this would possibly look something like this:
```
class MyCustomParticle(parcels.Particle):
    user_vars = {v1: np.int32, v2: np.int32}

    def sinking(particle, grid, time, dt):
        # Code that implements sinking

k_adv = parcels.Kernel(parcels.AdvectionRK4)
k_sink = parcels.Kernel(MyCustomKernel.sinking)
particleset.execute(k_adv + k_sink, ...)
```

The `k_adv + k_sink` operation would then cause the kernel objects to merge the inner code in Python as well as C/JIT mode and execute a single loop. This could become very powerful, since we could arbitrarily chain up small kernel functions, ranging from passive to active motion, as well as custom behaviour, eg. fish physiology. Does this sound like a practical and easy to understand API?

**Comment by erikvansebille** [comment]
Yes, I like this idea for an API. The only thing I think we need to make sure is possible within the PAI is for kernels to be run *conditionally*. For instance, for the sinking particles we want to stop advection when a particle hits the ocean floor. The results from one kernel can therefore be a condition to execute another kernel. We'll need to thing how this works within the API.

**Comment by mlange05** [comment]
Good point. Although we currently don't support if-conditionals in the custom JIT kernels we can easily add them. There is a whole bunch of Python language constructs that we currently don't deal with but they will be added gradually as we require them. For the sinking case an outer conditional should be enough to change behaviour depending on depth.

Calling a kernel from within a kernel, however, is the more interesting case and we need to think carefully about how to do this in the generated C code (it's reasonably easy in pure Python). We should be able to simply call (sub-)kernels, since they are all currently declared as `static inline`. This way we let the compiler do the work, instead of having to manually merge the code ASTs. Note though that this only works as long as the grid and particle type are the same for all sub-kernels.

**Comment by erikvansebille** [comment]
Implemented in #183
